csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -186,176 +156,104 @@ public CompressionResult CompressPayload()
-                            {
-                                Data.Add(defaultBinaryContext.Slice(startPos, istr.Pos - startPos));
-                                AddKey(key);
-                            }
-                        }
-                    }
+            // First write the overrides, then the InitialBinaryContext entries that were not overridden.
 
-                    ostr.RewriteFixedLengthSize20(_binaryContextKeys.Count, PayloadEnd, 2);
+            if (_binaryContextOverride is Dictionary<int, Action<OutputStream>> binaryContextOverride)
+            {
+                foreach ((int key, Action<OutputStream> action) in binaryContextOverride)
+                {
+                    ostr.WriteVarInt(key);
+                    OutputStream.Position startValue = ostr.StartFixedLengthSize(2);
+                    action(ostr);
+                    ostr.EndFixedLengthSize(startValue, 2);
+                    size++;
                 }
-                else
+            }
+            foreach ((int key, ReadOnlyMemory<byte> value) in InitialBinaryContext)
+            {
+                if (_binaryContextOverride == null || !_binaryContextOverride.ContainsKey(key))
                 {
-                    Debug.Assert(_binaryContextKeys == null);
-
-                    // Only when forwarding an ice2 request or response
-                    if (defaultBinaryContext.Count > 0 && Data[^1].Array == defaultBinaryContext.Array)
-                    {
-                        // Just expand the last segment to include the binary context bytes as-is.
-                        Data[^1] = new ArraySegment<byte>(Data[^1].Array!,
-                                                          Data[^1].Offset,
-                                                          Data[^1].Count + defaultBinaryContext.Count);
-                    }
-                    else
-                    {
-                        Data[^1] = Data[^1].Slice(0, PayloadEnd.Offset);
-                        if (defaultBinaryContext.Count > 0)
-                        {
-                            // Can happen if an interceptor compresses the payload
-                            Data.Add(defaultBinaryContext);
-                        }
-                    }
+                    ostr.WriteVarInt(key);
+                    ostr.WriteSize(value.Length);
+                    ostr.WriteByteSpan(value.Span);
+                    size++;
                 }
-
-                Size = Data.GetByteCount();
-                IsSealed = true;
             }
+            ostr.RewriteFixedLengthSize20(size, start, sizeLength);
         }
+    }
 
-        private protected bool ContainsKey(int key) => _binaryContextKeys?.Contains(key) ?? false;
-
-        private bool AddKey(int key)
-        {
-            _binaryContextKeys ??= new HashSet<int>();
-            return _binaryContextKeys.Add(key);
-        }
+    public static class OutgoingFrameHelper
```

Comment:
What is the purpose of this public method? It's not clear how users will be able to use it and for what purpose.

Reply:
This public extension method allows users to marshal the header of a frame.
I'll remove it for now, we can add it back if we find a use-case.